### ISSUE TITLE
Performance degradation b/w 7.4 and 7.10 for default flags?

### ISSUE THREAD

**Issue by sharatsc** [JohnLangford/vowpal_wabbit#602]
The benchmark uses 15000 examples from MNIST scaled and formatted as VW input. 
a) ECT
Training:
.\vw.7.4 --ect 26 -d .\letter.scale.vw --holdout_off -f letter.model.7.4
.\vw.7.10 --ect 26 -d .\letter.scale.vw --holdout_off -f letter.model.7.10
Prediction:
.\vw.7.4 -d .\letter.scale.vw -i .\letter.model.7.4 --testonly -l 0 -p predictions.testing.7.4
.\vw.7.10 -d .\letter.scale.vw -i .\letter.model.7.10 --testonly -l 0 -p predictions.testing.7.10

Evaluation (in R)
cut -f 1 -d " " .\letter.scale.vw > training.txt
training <- read.csv("training.txt", header=F)
testing.7.4 <- read.csv('predictions.testing.7.4', header=F)
testing.7.10 <- read.csv('predictions.testing.7.10', header=F)
mean(testing.7.10 == training)
[1] 0.3475333
mean(testing.7.4 == training)
[1] 0.4348
Accuracy has reduced from 43.48% to 34.7%

b) OOA
.\vw.7.4 --oaa 26 -d .\letter.scale.vw --holdout_off -f letter.model.7.4
.\vw.7.10 --oaa 26 -d .\letter.scale.vw --holdout_off -f letter.model.7.10

.\vw.7.4 -d .\letter.scale.vw -i .\letter.model.7.4 --testonly -l 0 -p predictions.testing.7.4
.\vw.7.10 -d .\letter.scale.vw -i .\letter.model.7.10 --testonly -l 0 -p predictions.testing.7.10

Evaluation: (R)
testing.7.4 <- read.csv('predictions.testing.7.4', header=F)
testing.7.10 <- read.csv('predictions.testing.7.10', header=F)
mean(testing.7.4 == training)
[1] 0.6472667
mean(testing.7.10 == training)
[1] 0.487

Accuracy has reduced from 64.72% to 48.7%

**Comment by sharatsc** [comment]
Looks like the drop in accuracy is restricted to simple --sgd update rule. Should we make --normalized default to avoid stumping invocations with default arguments?

Here are the performance numbers for different update rules (will use OAA only)
# SGD
## SGD 7.4
.\vw.7.4 --oaa 26 -d .\letter.scale.vw --sgd --holdout_off -f letter.model.7.4.sgd
.\vw.7.4 -d .\letter.scale.vw --testonly -i letter.model.7.4.sgd -p .\predictions.testing.7.4.sgd

>mean(testing.7.4.sgd == training): **0.6174667**

## SGD 7.10
.\vw.7.10 --oaa 26 -d .\letter.scale.vw --sgd --holdout_off -f letter.model.7.10.sgd
.\vw.7.10 -d .\letter.scale.vw --testonly -i letter.model.7.10.sgd -p .\predictions.testing.7.10.sgd
> mean(testing.7.10.sgd == training) [1]: **0.5872**

# Normalized
## Normalized 7.4
.\vw.7.4 --oaa 26 -d .\letter.scale.vw --normalized --holdout_off -f letter.model.7.4.normalized
.\vw.7.4 -d .\letter.scale.vw --testonly -i letter.model.7.4.normalized -p .\predictions.testing.7.4.normalized
>mean(testing.7.4.normalized == training): **0.6398667**


## Normalized 7.10
.\vw.7.10 --oaa 26 -d .\letter.scale.vw --normalized --holdout_off -f letter.model.7.10.normalized
.\vw.7.10 -d .\letter.scale.vw --testonly -i letter.model.7.10.normalized -p .\predictions.testing.7.10.normalized
> mean(testing.7.10.normalized == training): ** 0.6538 (improves)**

**Comment by EmergentOrder** [comment]
normalized is already on by default ( https://github.com/JohnLangford/vowpal_wabbit/wiki/Command-line-arguments ). Which is likely why this regression wasn't noticed / prioritized.

**Comment by sharatsc** [comment]
I think the documentation might be **wrong**?
## Default flags
.\vw.7.10 --oaa 26 -d .\letter.scale.vw --holdout_off -f letter.model.7.10.default
.\vw.7.10 -d .\letter.scale.vw --testonly -i letter.model.7.10.default -p .\predictions.testing.7.10.default

## with normalize
 .\vw.7.10 --oaa 26 -d .\letter.scale.vw --normalized --holdout_off -f letter.model.7.10.normalized
.\vw.7.10 -d .\letter.scale.vw --testonly -i letter.model.7.10.normalized -p .\predictions.testing.7.10.normalized

> mean(testing.7.10.default == training)
[1] 0.487
> mean(testing.7.10.normalized == training)
[1] 0.6538

**Comment by EmergentOrder** [comment]
Not only --normalized, but also --adaptive and --invariant are on by default.
if you specify --normalized though, only get normalized, not the other two. 
If you add all three, you should see the same as without any of the three.

**Comment by sharatsc** [comment]
You're right. I see the same **degradation** in accuracy by specifying --normalized --adaptive --invariant. We probably do not want a less accurate setting for default

**Comment by JohnLangford** [comment]
Something is off.  When I run:
vw --oaa 26 letter.scale.vw -f model; vw -i model letter.scale.vw -t
I see:
average loss = 0.329533
which indicates accuracy 0.670467

Maybe there is a windows/linux bug?  Or maybe evaluation is somehow screwed up?  Can you tell me what the average loss you see is?

BTW, 'letter' is different from 'mnist'.

**Comment by sharatsc** [comment]
You're right. I should have mentioned it was the letter dataset. Here's my run from head on Ubuntu 14.04

> sharat@ubuntu:~/vowpal_wabbit/vowpalwabbit$ ./vw -d ~/Desktop/letter.scale.vw --oaa 26 -f vw.model; ./vw -d ~/Desktop/letter.scale.vw -i vw.model -t 

final_regressor = vw.model
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = /home/sharat/Desktop/letter.scale.vw
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
1.000000 1.000000            1            1.0       26        1       17
1.000000 1.000000            2            2.0       16       26       17
1.000000 1.000000            4            4.0        8       19       17
0.875000 0.750000            8            8.0       18       14       17
0.937500 1.000000           16           16.0        8       19       17
0.937500 0.937500           32           32.0        4        1       17
0.937500 0.937500           64           64.0       26       15       17
0.945312 0.953125          128          128.0        9       23       17
0.921875 0.898438          256          256.0       10        7       17
0.896484 0.871094          512          512.0        4        7       17
0.823242 0.750000         1024         1024.0        1       12       17
0.770996 0.718750         2048         2048.0        1        1       17
0.713135 0.655273         4096         4096.0       25       17       17
0.655884 0.598633         8192         8192.0       15        7       17

finished run
number of examples per pass = 15000
passes used = 1
weighted example sum = 15000.000000
weighted label sum = 0.000000
average loss = 0.618600
total feature number = 255000
only testing
Num weight bits = 18
learning rate = 10
initial_t = 1
power_t = 0.5
using no cache
Reading datafile = /home/sharat/Desktop/letter.scale.vw
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
0.000000 0.000000            1            1.0       26       26       17
0.500000 1.000000            2            2.0       16        7       17
0.500000 0.500000            4            4.0        8        8       17
0.375000 0.250000            8            8.0       18       18       17
0.312500 0.250000           16           16.0        8        8       17
0.406250 0.500000           32           32.0        4        4       17
0.437500 0.468750           64           64.0       26        4       17
0.468750 0.500000          128          128.0        9        9       17
0.500000 0.531250          256          256.0       10        3       17
0.515625 0.531250          512          512.0        4        4       17
0.522461 0.529297         1024         1024.0        1        1       17
0.521484 0.520508         2048         2048.0        1        1       17
0.523193 0.524902         4096         4096.0       25        1       17
0.514893 0.506592         8192         8192.0       15        8       17

finished run
number of examples per pass = 15000
passes used = 1
weighted example sum = 15000.000000
weighted label sum = 0.000000
average loss = 0.513000
total feature number = 255000

**Comment by JohnLangford** [comment]
So far, this looks like a windows/Linux bug.  I'm investigating.

**Comment by JohnLangford** [comment]
This was a bug due to SSE2 not working properly on windows leading to
1/sqrt(x) effectively becoming 1/x.

https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/gd.cc#L83

SSE2 is still not working properly, but it's at least computing something
like 1/sqrt(x).

-John


On Tue, Apr 28, 2015 at 10:14 AM, Sharat Chikkerur <notifications@github.com
